zebra-chain/src/parameters/network.rs

```diff
@@ -53,21 +50,109 @@ const ZIP_212_GRACE_PERIOD_DURATION: HeightDiff = 32_256;
 
 /// An enum describing the possible network choices.
 #[derive(Copy, Clone, Debug, Default, Eq, PartialEq, Hash, Serialize, Deserialize)]
-#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
 pub enum Network {
     /// The production mainnet.
     #[default]
     Mainnet,
 
     /// The oldest public test network.
-    Testnet,
+    Testnet(#[serde(skip)] TestnetParameters),
+}
+
+/// Configures testnet network parameters to use instead of default values.
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq, Hash, Serialize)]
+pub struct TestnetParameters(Option<NetworkParameters>);
+
+impl std::ops::Deref for TestnetParameters {
+    type Target = Option<NetworkParameters>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<Option<NetworkParameters>> for TestnetParameters {
+    fn from(value: Option<NetworkParameters>) -> Self {
+        Self(value)
+    }
+}
+
+impl TestnetParameters {
+    /// Returns `network_id` in network parameters, if any.
+    pub fn network_id(self) -> Option<[u8; 4]> {
+        self.0.and_then(NetworkParameters::network_id)
+    }
+
+    /// Returns `default_port` in network parameters, if any.
+    pub fn default_port(self) -> Option<u16> {
+        self.0.and_then(NetworkParameters::default_port)
+    }
+
+    /// Returns `cache_name` in network parameters, if any.
+    pub fn cache_name(self) -> Option<&'static str> {
+        self.0.and_then(NetworkParameters::cache_name)
+    }
+
+    /// Returns `genesis_hash` in network parameters, if any.
+    pub fn genesis_hash(self) -> Option<&'static str> {
+        self.0.and_then(NetworkParameters::genesis_hash)
+    }
+
+    /// Returns `activation_heights` in network parameters, if any.
+    pub fn activation_heights(self) -> Option<&'static [(Height, NetworkUpgrade)]> {
+        self.0.and_then(NetworkParameters::activation_heights)
+    }
+}
+
+/// Configures network parameters to use instead of default values.
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq, Hash, Serialize)]
+pub struct NetworkParameters {
+    /// Used as the network magic, Zebra will reject messages and connections from peers
+    /// with a different network magic.
+    network_id: Option<[u8; 4]>,
+
+    /// Default port for the network
+    default_port: Option<u16>,
+
+    /// Network portion of zebra-state cache path
+    cache_name: Option<&'static str>,
+
+    /// Genesis hash for this testnet
+    genesis_hash: Option<&'static str>,
+
+    /// Activation heights for this testnet
+    activation_heights: Option<&'static [(Height, NetworkUpgrade)]>,
+}
+
+impl NetworkParameters {
+    fn network_id(self) -> Option<[u8; 4]> {
+        self.network_id
+    }
```

Comment:
I think this might be a misunderstanding of how the Zcash network protocol works, or I might have used the wrong words to describe it.

What you're looking for is the network magic number:
https://github.com/ZcashFoundation/zebra/blob/2efc6bf4b2875f16a987bcbff0f90cff98eda3fa/zebra-network/src/constants.rs#L418-L425

Which is checked in every network message's header:
https://github.com/ZcashFoundation/zebra/blob/2efc6bf4b2875f16a987bcbff0f90cff98eda3fa/zebra-network/src/protocol/external/codec.rs#L166
https://github.com/ZcashFoundation/zebra/blob/2efc6bf4b2875f16a987bcbff0f90cff98eda3fa/zebra-network/src/protocol/external/codec.rs#L392-L394

And you'll need to replace this code with the relevant method:
https://github.com/ZcashFoundation/zebra/blob/2efc6bf4b2875f16a987bcbff0f90cff98eda3fa/zebra-network/src/protocol/external/types.rs#L27-L35

This should be documented here, but it hasn't been documented yet:
https://github.com/zcash/zips/blob/b1410c0f42c0f08429046309393038bb6f275113/zip-0204.rst#L4

Reply:
I wanted to leave placeholders and didn't look at it close enough on the first pass. Fixed in https://github.com/ZcashFoundation/zebra/pull/7924/commits/8fa3a533283d8043ce01d4d690af6cdc4b4eb8db.